tango/executors/multicore_executor.py

```diff
@@ -0,0 +1,237 @@
+import logging
+import os
+import subprocess
+import time
+from tempfile import NamedTemporaryFile
+from typing import Dict, List, Optional, OrderedDict, Set, TypeVar
+
+from tango.executor import Executor
+from tango.step import Step
+from tango.step_graph import StepGraph
+from tango.workspace import StepState, Workspace
+
+logger = logging.getLogger(__name__)
+
+T = TypeVar("T")
+
+
+class MulticoreExecutor(Executor):
+    """
+    A ``MulticoreExecutor`` runs the steps in parallel and caches their results.
+    """
+
+    def __init__(
+        self,
+        workspace: Workspace,
+        include_package: Optional[List[str]] = None,
+        parallelism: int = 1,
+        num_tries_to_sync_states: int = 3,
+        wait_seconds_to_sync_states: int = 3,
+    ) -> None:
+        self.workspace = workspace
+        self.include_package = include_package
+        self.parallelism = parallelism
+
+        # Perhaps there's a better way to do this without these being passed as args.
+        self._num_tries_to_sync_states = num_tries_to_sync_states
+        self._wait_seconds_to_sync_states = wait_seconds_to_sync_states
+
+    def execute_step_graph(self, step_graph: StepGraph, run_name: Optional[str] = None) -> Set[str]:
+        """
+        Execute a :class:`tango.step_graph.StepGraph`.
```

Comment:
I have implemented it according to policy 2. If something fails, we don't run anything that requires that step, but we _do_ run unrelated steps. This makes sense to me because say, I kick off some long processing steps overnight, and the first one fails -- it would be nice to have the remaining ones still finished by the morning. I will add this to the docstring, but let me know if you are strongly in favor of policy 1.

Reply:
No, makes sense to me. Policy 2 it is!